// @toverux/expresse/dist/src/redis_hub.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Redis = require("ioredis");
const hub_1 = require("./hub");
class RedisHub extends hub_1.Hub {
    constructor(channel, pub = new Redis(), sub = new Redis()) {
        super();
        this.channel = channel;
        this.pub = pub;
        this.sub = sub;
        sub.subscribe(channel);
        sub.on('message', (chan, message) => this.handleRedisMessage(message));
    }
    data(data, id) {
        this.publish('data', arguments);
    }
    event(event, data, id) {
        this.publish('event', arguments);
    }
    comment(comment) {
        this.publish('comment', arguments);
    }
    publish(type, args) {
        const message = {
            type, args: Array.from(args)
        };
        this.pub.publish(this.channel, JSON.stringify(message));
    }
    handleRedisMessage(jsonMessage) {
        const { type, args } = JSON.parse(jsonMessage);
        super[type].apply(this, args);
    }
}
exports.RedisHub = RedisHub;
